Rewrite the PhenoDigm parser in PySpark

tskir opened the pull request:
Fixes https://github.com/opentargets/platform/issues/1403.

### Finalised discussions (kept for historical purposes)

* [x] **Ontology mapping.** When computing `diseaseFromSourceId`, the original format is used without any transformations. For example, `ORPHA:3085` stays as is and does **not** become `Orphanet:3085` or `Orphanet_3085`.
* [x] **Gene mapping.** Despite this not being strictly speaking necessary anymore, the parser still performs the mapping of HGNC IDs to human Ensembl gene IDs, and outputs them in the `targetFromSourceId` field.
* [x] **Phenotypes.** The phenotypes are processed according to the logic of the original script:
  + `diseaseModelAssociatedModelPhenotypes` include **all** phenotypes of the mouse model, not taking any disease information into account. This is to give the complete representation of the original model.
  + `diseaseModelAssociatedHumanPhenotypes` contain only the phenotypes which are both present in the disease, **and** also could be traced back to the mouse model phenotypes via the ontology mapping table. This is to better represent the translated disease-target associations.
* [x] **Models, diseases, and their associations.**
  + For non-standard model IDs (`unreleased-*`), we nullify that field (but do not remove the evidence).
  + As per the original script, when computing `biologicalModelId`, everything after `#` is removed. For example, `MGI:6274930#hom#early` becomes `MGI:6274930`. However, this is fine because the `hom early` part of the information is duplicated in `biologicalModelGeneticBackground` anyway.
  + The original script used to filter the disease–model associations on the `disease_model_max_norm` score, with a 90 out of 100 threshold by default. In line with other evidence sources, this has now been disabled, and all associations are being output.

### Related tickets
* [x] An old issue https://github.com/opentargets/platform/issues/859 about the allele IDs not being populated. The field is present in the new evidence strings and the issue is resolved.
* [x] A more recent issue https://github.com/opentargets/platform/issues/1473 about constructing the links to MGI models. Reported by the front-end team as done.

### 10 randomly chosen evidence string examples from the new pipeline
```json
{"biologicalModelAllelicComposition":"Fgfr1<tm2Jrt>/Fgfr1<tm2Jrt>","biologicalModelGeneticBackground":"involves: 129S1/Sv * 129X1/SvJ * ICR","biologicalModelId":"MGI:3702995","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Short Stature-Craniofacial Anomalies-Genital Hypoplasia Syndrome","diseaseFromSourceId":"ORPHA:2994","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0000368","label":"Low-set, posteriorly rotated ears"},{"id":"HP:0008551","label":"Microtia"},{"id":"HP:0000175","label":"Cleft palate"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0003235","label":"abnormal alisphenoid bone morphology"},{"id":"MP:0000111","label":"cleft palate"},{"id":"MP:0000762","label":"abnormal tongue morphology"},{"id":"MP:0030119","label":"abnormal retrotympanic process morphology"},{"id":"MP:0005106","label":"abnormal incus morphology"},{"id":"MP:0011087","label":"neonatal lethality, complete penetrance"},{"id":"MP:0002950","label":"abnormal neural crest cell migration"},{"id":"MP:0003138","label":"absent tympanic ring"},{"id":"MP:0000040","label":"absent middle ear ossicles"},{"id":"MP:0000042","label":"abnormal organ of Corti morphology"},{"id":"MP:0009655","label":"abnormal secondary palate development"},{"id":"MP:0008381","label":"absent gonial bone"},{"id":"MP:0005105","label":"abnormal middle ear ossicle morphology"},{"id":"MP:0004453","label":"abnormal pterygoid bone morphology"},{"id":"MP:0006338","label":"abnormal second pharyngeal arch morphology"},{"id":"MP:0030125","label":"small gonial bone"},{"id":"MP:0006020","label":"decreased tympanic ring size"},{"id":"MP:0009845","label":"abnormal neural crest cell morphology"},{"id":"MP:0004402","label":"decreased cochlear outer hair cell number"},{"id":"MP:0009886","label":"failure of palatal shelf elevation"},{"id":"MP:0000029","label":"abnormal malleus morphology"},{"id":"MP:0009914","label":"abnormal hyoid bone lesser horn morphology"},{"id":"MP:0003308","label":"abnormal cochlear sensory epithelium morphology"},{"id":"MP:0000018","label":"small ears"},{"id":"MP:0008023","label":"abnormal styloid process morphology"}],"resourceScore":0.9584999999999999,"targetFromSourceId":"ENSG00000077782","targetInModel":"Fgfr1","targetInModelId":"ENSMUSG00000031565"}
{"biologicalModelAllelicComposition":"Lrp4<dan>/Lrp4<dan>","biologicalModelGeneticBackground":"129S2/SvPas-Lrp4<dan>","biologicalModelId":"MGI:3623728","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Gombo Syndrome","diseaseFromSourceId":"OMIM:233270","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0009466","label":"Radial deviation of finger"},{"id":"HP:0001156","label":"Brachydactyly"},{"id":"HP:0030084","label":"Clinodactyly"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0000562","label":"polydactyly"},{"id":"MP:0004083","label":"polysyndactyly"},{"id":"MP:0001510","label":"abnormal coat appearance"},{"id":"MP:0001382","label":"abnormal nursing"},{"id":"MP:0002544","label":"brachydactyly"},{"id":"MP:0000580","label":"deformed nails"},{"id":"MP:0001265","label":"decreased body size"},{"id":"MP:0000564","label":"syndactyly"},{"id":"MP:0001505","label":"hunched posture"}],"resourceScore":1.0,"targetFromSourceId":"ENSG00000134569","targetInModel":"Lrp4","targetInModelId":"ENSMUSG00000027253"}
{"biologicalModelAllelicComposition":"Bmp5<se-4J>/Bmp5<se-4J>","biologicalModelGeneticBackground":"B6(Cg)-Bmp5<se-4J>/J","biologicalModelId":"MGI:3764963","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Mental Retardation, Autosomal Recessive 35","diseaseFromSourceId":"OMIM:615162","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0008551","label":"Microtia"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0002177","label":"abnormal outer ear morphology"},{"id":"MP:0002102","label":"abnormal ear morphology"},{"id":"MP:0000022","label":"abnormal ear shape"},{"id":"MP:0000018","label":"small ears"}],"resourceScore":0.9331999999999999,"targetFromSourceId":"ENSG00000112175","targetInModel":"Bmp5","targetInModelId":"ENSMUSG00000032179"}
{"biologicalModelAllelicComposition":"Zfp60<tm1b(KOMP)Mbp> het early","biologicalModelGeneticBackground":"C57BL/6NJ","biologicalModelId":"MGI:5635493","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"X-Linked Intellectual Disability, Stocco Dos Santos Type","diseaseFromSourceId":"ORPHA:85288","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0000752","label":"Hyperactivity"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0020870","label":"decreased thigmotaxis"},{"id":"MP:0001399","label":"hyperactivity"}],"resourceScore":0.9146,"targetFromSourceId":"ENSG00000197782","targetInModel":"Zfp60","targetInModelId":"ENSMUSG00000037640"}
{"biologicalModelAllelicComposition":"Fgfr2<tm3.1Cxd>/Fgfr2<+>","biologicalModelGeneticBackground":"involves: 129S6/SvEvTac * FVB/N","biologicalModelId":"MGI:6407058","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Radiohumeral Fusions With Other Skeletal And Craniofacial Anomalies","diseaseFromSourceId":"OMIM:614416","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0001166","label":"Arachnodactyly"},{"id":"HP:0001363","label":"Craniosynostosis"},{"id":"HP:0000248","label":"Brachycephaly"},{"id":"HP:0012165","label":"Oligodactyly"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0003662","label":"abnormal long bone epiphyseal plate proliferative zone"},{"id":"MP:0002083","label":"premature death"},{"id":"MP:0012085","label":"midface hypoplasia"},{"id":"MP:0012173","label":"short rostral-caudal axis"},{"id":"MP:0001300","label":"ocular hypertelorism"},{"id":"MP:0001265","label":"decreased body size"},{"id":"MP:0030420","label":"short basicranium"},{"id":"MP:0008272","label":"abnormal endochondral bone ossification"},{"id":"MP:0001262","label":"decreased body weight"},{"id":"MP:0011861","label":"increased cranium height"},{"id":"MP:0010743","label":"delayed cranial suture closure"},{"id":"MP:0001732","label":"postnatal growth retardation"},{"id":"MP:0000564","label":"syndactyly"},{"id":"MP:0000131","label":"abnormal long bone epiphysis morphology"},{"id":"MP:0030095","label":"abnormal midface morphology"},{"id":"MP:0030467","label":"abnormal cranial cartilage development"},{"id":"MP:0000120","label":"malocclusion"},{"id":"MP:0011086","label":"postnatal lethality, incomplete penetrance"},{"id":"MP:0011710","label":"enhanced osteoblast differentiation"},{"id":"MP:0000438","label":"abnormal cranium morphology"},{"id":"MP:0030350","label":"premature coronal suture closure"},{"id":"MP:0006399","label":"abnormal long bone epiphyseal ossification zone morphology"},{"id":"MP:0030443","label":"abnormal cranial synchondrosis"},{"id":"MP:0011862","label":"decreased cranium length"},{"id":"MP:0000440","label":"domed cranium"},{"id":"MP:0000435","label":"shortened head"},{"id":"MP:0000428","label":"abnormal craniofacial morphology"},{"id":"MP:0000130","label":"abnormal trabecular bone morphology"},{"id":"MP:0003409","label":"decreased width of hypertrophic chondrocyte zone"},{"id":"MP:0004053","label":"abnormal synchondrosis"},{"id":"MP:0003418","label":"premature intramembranous bone ossification"},{"id":"MP:0000441","label":"increased cranium width"}],"resourceScore":0.9039,"targetFromSourceId":"ENSG00000066468","targetInModel":"Fgfr2","targetInModelId":"ENSMUSG00000030849"}
{"biologicalModelAllelicComposition":"Nanos3<tm1Ysa>/Nanos3<tm1Ysa>","biologicalModelGeneticBackground":"involves: 129S1/Sv * 129X1/SvJ * ICR","biologicalModelId":"MGI:2682627","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Spermatogenic Failure 32","diseaseFromSourceId":"OMIM:618115","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0000789","label":"Infertility"},{"id":"HP:0000027","label":"Azoospermia"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0001925","label":"male infertility"},{"id":"MP:0005159","label":"azoospermia"},{"id":"MP:0004806","label":"absent germ cells"},{"id":"MP:0004852","label":"decreased testis weight"},{"id":"MP:0001116","label":"small gonad"},{"id":"MP:0001926","label":"female infertility"},{"id":"MP:0008392","label":"decreased primordial germ cell number"},{"id":"MP:0001127","label":"small ovary"}],"resourceScore":1.0,"targetFromSourceId":"ENSG00000187556","targetInModel":"Nanos3","targetInModelId":"ENSMUSG00000056155"}
{"biologicalModelAllelicComposition":"Fmr1<tm1.1Cidz>/Y","biologicalModelGeneticBackground":"involves: 129P2/OlaHsd * C57BL/6","biologicalModelId":"MGI:3808886","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Mental Retardation, X-Linked 2","diseaseFromSourceId":"OMIM:300428","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0000053","label":"Macroorchidism"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0001148","label":"enlarged testis"}],"resourceScore":1.0,"targetFromSourceId":"ENSG00000102081","targetInModel":"Fmr1","targetInModelId":"ENSMUSG00000000838"}
{"biologicalModelAllelicComposition":"Tyr<c-18J>/Tyr<c-18J>","biologicalModelGeneticBackground":"C57BL/6J H2<bm3>/EgJ-Tyr<c-18J>","biologicalModelId":"MGI:3819166","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Spastic Paraplegia-Facial-Cutaneous Lesions Syndrome","diseaseFromSourceId":"ORPHA:2819","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0000953","label":"Hyperpigmentation of the skin"},{"id":"HP:0001053","label":"Hypopigmented skin patches"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0005171","label":"absent coat pigmentation"},{"id":"MP:0001189","label":"absent skin pigmentation"}],"resourceScore":0.9661,"targetFromSourceId":"ENSG00000077498","targetInModel":"Tyr","targetInModelId":"ENSMUSG00000004651"}
{"biologicalModelAllelicComposition":"Myo5a<d-l32J>/Myo5a<d-l32J>","biologicalModelGeneticBackground":"B6.Cg-Myo5a<d-l32J>/GrsrJ","biologicalModelId":"MGI:4940035","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Molybdenum Cofactor Deficiency, Complementation Group C","diseaseFromSourceId":"OMIM:615501","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0002179","label":"Opisthotonus"},{"id":"HP:0002197","label":"Generalized-onset seizure"},{"id":"HP:0002069","label":"Bilateral tonic-clonic seizure"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0000371","label":"diluted coat color"},{"id":"MP:0003996","label":"clonic seizures"},{"id":"MP:0002083","label":"premature death"},{"id":"MP:0002880","label":"opisthotonus"},{"id":"MP:0001525","label":"impaired balance"},{"id":"MP:0011086","label":"postnatal lethality, incomplete penetrance"},{"id":"MP:0001393","label":"ataxia"}],"resourceScore":0.9473999999999999,"targetFromSourceId":"ENSG00000197535","targetInModel":"Myo5a","targetInModelId":"ENSMUSG00000034593"}
{"biologicalModelAllelicComposition":"Ntrk3<tm1Bbd>/Ntrk3<tm1Bbd>","biologicalModelGeneticBackground":"involves: 129S2/SvPas * C57BL/6","biologicalModelId":"MGI:2175190","datasourceId":"phenodigm","datatypeId":"animal_model","diseaseFromSource":"Diabetes Mellitus, Permanent Neonatal 3","diseaseFromSourceId":"OMIM:618857","diseaseModelAssociatedHumanPhenotypes":[{"id":"HP:0002305","label":"Athetosis"}],"diseaseModelAssociatedModelPhenotypes":[{"id":"MP:0003986","label":"small cochlear ganglion"},{"id":"MP:0001527","label":"athetotic walking movements"},{"id":"MP:0002082","label":"postnatal lethality"},{"id":"MP:0002083","label":"premature death"},{"id":"MP:0003987","label":"small vestibular ganglion"},{"id":"MP:0000968","label":"abnormal sensory neuron innervation pattern"},{"id":"MP:0006404","label":"abnormal lumbar dorsal root ganglion morphology"},{"id":"MP:0001265","label":"decreased body size"},{"id":"MP:0004746","label":"abnormal cochlear IHC afferent innervation pattern"},{"id":"MP:0001405","label":"impaired coordination"},{"id":"MP:0003308","label":"abnormal cochlear sensory epithelium morphology"},{"id":"MP:0001524","label":"impaired limb coordination"},{"id":"MP:0001513","label":"limb grasping"}],"resourceScore":0.9606999999999999,"targetFromSourceId":"ENSG00000140538","targetInModel":"Ntrk3","targetInModelId":"ENSMUSG00000059146"}
```

### Points for the follow-up issue
* **General improvements**
  + [ ] A nice addition to the script would be to calculate the various quality control metrics. For example, which percentage of the original disease-model associations were eventually converted into the evidence strings, and what are the reasons for the failing records.
  + [ ] The static MGI and HGNC mapping files could be downloaded using Snakemake (once that PR is merged) instead of directly by the script.
  + [ ] Conduct more detailed investigations on the meaning of `disease_model_max_norm`, perhaps by reading the PhenoDigm paper or contacting the developers. Decide whether to filter on it or not.
* **Representation improvements**
  + [ ] For the 21.06 release, we must ensure that the platform backend & frontend are updated to reflect the updated evidence schema regarding mouse gene ID and name.
  + [ ] The platform website could/should provide more information on how the mouse model data are computed and what is the meaning of different fields. In particular, which phenotypes are presented.
  + [ ] It might be useful to get an input from someone who is working with mouse models a lot, to see how people are using this information; if they are satistifed; if they want additional fields, etc.
* **Ontology mappings**
  + [ ] As per discussions, `diseaseFromSourceMappedId` is currently **not** populated by the refactored pipeline. It is expected that this field is going to be populated by downstream pipelines. If this functionality is still not available in time for the 21.06 release, this functionality will have to be temporarily implemented (again) in the parser.

ireneisdoomed commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+                                                rows=IMPC_SOLR_BATCH_SIZE)
+        for data_type in IMPC_SOLR_TABLES:
+            self.logger.info(f'Fetching PhenoDigm data type {data_type}.')
+            filename = os.path.join(self.cache_dir, IMPC_FILENAME.format(data_type=data_type))
```
Use f-strings instead.

tskir replied:
I'm formatting a constant here so it cannot be replaced by an f-string, unless I misunderstand something. Could you provide an example of what you're proposing please?

ireneisdoomed commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+
+    def load_solr_csv(self, data_type):
+        """Load the CSV from SOLR; rename and select columns as specified."""
+        df = self.spark.read.csv(os.path.join(self.cache_dir, IMPC_FILENAME.format(data_type=data_type)), header=True)
```
Use f-strings instead.

tskir replied:
Same as [above](https://github.com/opentargets/evidence_datasource_parsers/pull/80#discussion_r616877506) :)

ireneisdoomed commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    PhenoDigm(logging, cache_dir).process_all(output, score_cutoff, use_cached)
+
+
+if __name__ == '__main__':
```
I personally think it is a better practice to include this logic within main and in this if block just call the main() function.

tskir replied:
Ah, this is where I completely disagree, if you don't mind. To me, the entire point of having the `if __name__ == '__main__'` block is to be able to import the functions (including `main()` if you want to) from the module into another one.

If we define a parser inside `main()` and then try to import it, it will break things heavily, because the command line arguments for the parser won't be there anymore.

ireneisdoomed reviewed (changes requested):
Addressing your comments:
1. **Ontology mapping**. Good point on the `diseaseFromSourceMappedId`. My impression is that we have to stick to what the source reports in this case. This correction should come in the disease mapping step. 
2. **Gene mapping**. I would remove this step unless it is necessary in one of the joins of the script. In the case of the mice targets, it may be worth discussing this step in the future, but right now I don't think we use it in any other scripts.
3.1. **Disease–model associations.** Yes. `biologicalModelId` is strictly an ID and any other annotation wouldn't match the pattern.
3.2. How frequent is `disease_model_max_norm` there? Is this a legacy filter or is Phenodigm indicating that anything below this number lacks confidence? The overall approach is to collect all sorts of evidence, even the negative one. This is worth discussing.
3.4. Shouldn't we be describing the result of the mapping between `diseaseModelAssociatedModelPhenotypes` and `diseaseModelAssociatedHumanPhenotypes`?
3.4. With `diseaseModelAssociatedHumanPhenotypes `, what do you mean by the phenotypes present in the disease?

My comments in general is that I would review the caching feature as this does not seem to be necessary performance-wise. If we decide to remove it, the code would be much simpler and much more explicit as we would not have to reuse so many functions.

Then, as proposed changes and also as a consequence of the above, I would declare less objects using Python and would concatenate more in-built Spark functions instead. In addition, I would move the `process_all` logic to the `main()` function, as well as the logic inside the last if block.
We are approaching the logging systematically across all the parsers, so I would take a look at that. 

Awesome job anyway!

ireneisdoomed commented:
Addressing your comments:

Ontology mapping. Good point on the diseaseFromSourceMappedId. My impression is that we have to stick to what the source reports in this case. This correction should come in the disease mapping step.
Gene mapping. I would remove this step unless it is necessary in one of the joins of the script. In the case of the mice targets, it may be worth discussing this step in the future, but right now I don't think we use it in any other scripts.
3.1. Disease–model associations. Yes. biologicalModelId is strictly an ID and any other annotation wouldn't match the pattern.
3.2. How frequent is disease_model_max_norm there? Is this a legacy filter or is Phenodigm indicating that anything below this number lacks confidence? The overall approach is to collect all sorts of evidence, even the negative one. This is worth discussing.
3.4. Shouldn't we be describing the result of the mapping between diseaseModelAssociatedModelPhenotypes and diseaseModelAssociatedHumanPhenotypes?
3.4. With diseaseModelAssociatedHumanPhenotypes , what do you mean by the phenotypes present in the disease?
My comments in general is that I would review the caching feature as this does not seem to be necessary performance-wise. If we decide to remove it, the code would be much simpler and much more explicit as we would not have to reuse so many functions.

Then, as proposed changes and also as a consequence of the above, I would declare less objects using Python and would concatenate more in-built Spark functions instead. In addition, I would move the process_all logic to the main() function, as well as the logic inside the last if block.
We are approaching the logging systematically across all the parsers, so I would take a look at that.

Awesome job anyway!

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+from retry import retry
+
+
+# Human gene mappings.
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+
+# Human gene mappings.
+HGNC_DATASET_URI = 'http://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/tsv/hgnc_complete_set.txt'
+HGNC_DATASET_FILENAME = os.path.split(HGNC_DATASET_URI)[-1]
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+HGNC_DATASET_FILENAME = os.path.split(HGNC_DATASET_URI)[-1]
+
+# Mouse gene mappings.
+MGI_DATASET_URI = 'http://www.informatics.jax.org/downloads/reports/MGI_Gene_Model_Coord.rpt'
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+MGI_DATASET_FILENAME = os.path.split(MGI_DATASET_URI)[-1]
+
+# Mouse model data from IMPC SOLR.
+IMPC_SOLR_HOST = 'http://www.ebi.ac.uk/mi/impc/solr/phenodigm/select'
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+# The largest table is about 7 million records. The one billion limit is used as an arbitrary high number to retrieve
+# all records in one large request, which maximises the performance.
+IMPC_SOLR_BATCH_SIZE = 1000000000
+IMPC_SOLR_TIMEOUT = 3600
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    'ontology': ('ontology', 'phenotype_id', 'phenotype_term'),
+
+}
+IMPC_FILENAME = 'impc_solr_{data_type}.csv'
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    """Retrieve the data, load it into Spark, process and write the resulting evidence strings."""
+
+    def __init__(self, logger, cache_dir):
+        super(PhenoDigm, self).__init__()
```
Thank you very much for noticing! This `super()` call was a remnant from when this class indeed had multiple inheritance. Now it's not needed at all. Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/2368b34fcb3e0c3dd7b4325294287d4e555e2f78

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+                with open(os.path.join(tmp_dir_name, json_chunk_filename), 'rb') as json_chunk:
+                    shutil.copyfileobj(json_chunk, outfile)
+
+    def process_all(self, output, score_cutoff, use_cached):
```
I think it's fine either way :) I don't mind putting them into `main()`. Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/1d7cf1c69700d79909faae3d78627ff77540a17e

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    logging.basicConfig(**logging_config)
+
+    # Process the data.
+    PhenoDigm(logging, cache_dir).process_all(output, score_cutoff, use_cached)
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/1d7cf1c69700d79909faae3d78627ff77540a17e

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+
+# Mouse gene mappings.
+MGI_DATASET_URI = 'http://www.informatics.jax.org/downloads/reports/MGI_Gene_Model_Coord.rpt'
+MGI_DATASET_FILENAME = os.path.split(MGI_DATASET_URI)[-1]
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+# all records in one large request, which maximises the performance.
+IMPC_SOLR_BATCH_SIZE = 1000000000
+IMPC_SOLR_TIMEOUT = 3600
+DEFAULT_ASSOCIATION_SCORE_CUTOFF = 90.0
```
Thank you both for the suggestion! Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/b6f607002964221a55990d112f989d62e199e8c3

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+IMPC_FILENAME = 'impc_solr_{data_type}.csv'
+# The largest table is about 7 million records. The one billion limit is used as an arbitrary high number to retrieve
+# all records in one large request, which maximises the performance.
+IMPC_SOLR_BATCH_SIZE = 1000000000
```
Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/d86d627aaba9b73cc2148e829d18b4d23a8e2458

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+            filename = os.path.join(self.cache_dir, IMPC_FILENAME.format(data_type=data_type))
+            impc_solr_retriever.fetch_data(data_type, filename)
+
+    def load_tsv(self, filename):
```
I'm not against this suggestion, however, may I propose that we do this in a separate follow-up issue? The thing is, Snakemake isn't available in the main branch yet, and it seems to me that doing this _right now_ would make make the script harder not easier

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    def load_solr_csv(self, data_type):
+        """Load the CSV from SOLR; rename and select columns as specified."""
+        df = self.spark.read.csv(os.path.join(self.cache_dir, IMPC_FILENAME.format(data_type=data_type)), header=True)
+        column_name_mappings = [column_map.split(' > ') for column_map in IMPC_SOLR_TABLES[data_type]]
```
You know what, I concede that perhaps defining this custom renaming mechanism (the whole ` > ` thing) was rather unnecessary after all. It saves like 7 lines of code but admittedly makes things obscure as hell.

I've now made the column renames explicit and moved them to the place where each individual dataset is being loaded. Addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/bdd65c370974833ade8e3d0feb32810a2bd8c933

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+
+        # We are reporting all mouse phenotypes for a model, regardless of whether they can be mapped into any human
+        # disease.
+        all_mouse_phenotypes = (
```
The entire struct definition code is courtesy of @DSuveges, so all glory should go to him! :)

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+        # Split lists of phenotypes in the `mouse_model` and `disease` tables and keep only the ID. For example, one row
+        # with 'MP:0001529 abnormal vocalization,MP:0002981 increased liver weight' becomes two rows with 'MP:0001529'
+        # and 'MP:0002981'.
+        model_mouse_phenotypes = (
```
Thank you for a great suggestion, addressed in https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/a33c14012419d72a56140dcce863566741f822c6

tskir replied:
In the end I realised that `regexp_extract` doesn't allow the things to be sufficiently simplified, because it can only extract one match. I was able to solve it with upgrading to Spark 3.1.1 and using `expr(regexp_extract_all)`: https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/015cb611db024e84a61968ec167321fe8342e5d9. What do you think?

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+"""Evidence parser for the animal model sources from PhenoDigm."""
+
+import argparse
+import logging
```
I'm a bit confused, because as far as I can see my implementation already follows this convention: https://github.com/opentargets/evidence_datasource_parsers/pull/80/files#diff-2b946d8411b4433f2c4e427134659d102319c5317c1981bccc25d4972c32c9c2R323

The only thing it lacks is the else block, but this is only because the `stderr` stream handler is [enabled by default](https://docs.python.org/3/library/logging.handlers.html#logging.StreamHandler) and doesn't need to be specified.

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    def query_solr(self, data_type, start):
+        """Request one batch of SOLR records of the specified data type and write it into a temporary file."""
+        list_of_columns = [column.split(' > ')[0] for column in IMPC_SOLR_TABLES[data_type]]
+        params = {'q': '*:*', 'fq': f'type:{data_type}', 'start': start, 'rows': self.rows, 'wt': 'csv',
```
Thanks! I noticed that it gives quite a boost in download time, especially for `disease_model_summary` :)

tskir commented on modules/PhenoDigm.py:
```diff
@@ -0,0 +1,362 @@
+    def fetch_data(self, data_type, output_filename):
+        """Fetch all rows of the requested data type to the specified location."""
+        total_records = self.get_number_of_solr_records(data_type)
+        assert total_records != 0, f'SOLR did not return any data for {data_type}.'
```
This is just a passing sanity check: for any of the tables we're requesting, SOLR should not return an empty response (no rows). Hence this assert, which will throw an exception if this is not the case

tskir commented:
@DSuveges @ireneisdoomed All remaining investigations are now completed; ready for the second review iteration. Updates since the last time:
* Addressed all review comments
* For mouse model, process both gene ID and gene name
* Exclude non-standard model IDs (`NOT-RELEASED-*`)
* Investigated the model IDs and whether stripping the `#het#early` modifiers is justified (it is; added note to the relevant code step)

And most importantly, I disabled the score filtering by default. Unfortunately I wasn't able to find any explanations on what this score means (googling "disease_model_max_norm" literally only returns two hits, one of which is this PR). But since this is a normalised association score, I think we should be perfectly fine continue just running with it.

Upon investigation, it turns out that this score is always present, and is distributed like this:
![image](https://user-images.githubusercontent.com/10669118/117039488-be585f80-ad00-11eb-8c0e-0f12ff380833.png)

Previously, we had been excluding everything with the association score < 0.9, losing a huge chunk of all possible evidences. Given that the other parsers now return all evidences, and given that these records are all appropriately scored (or at least it looks like it), I think we should also process all of them in PhenoDigm.

DSuveges reviewed (approved):
I absolutely love this new version of the parser.

ireneisdoomed reviewed (dismissed):
Thanks for addressing everything! It looks great and ready to be merged, I only have some open questions:
1. Should we be normalising the score ourselves? The current approach is to move this step to the ETL.
2. Why are you doing the assertion in write_evidence_strings? You are already giving instructions to coalesce all chunks into one, in which scenario this could be different?
3. Regarding the scoring: In Phenodigm the scores report the association between diseases and animal models, not genes. The phenotype similarity is computed using an algorithm called OWLSim which expresses the similarity in terms of the Jaccard Index (simJ) or Information Content (IC). Therefore, to compute the score you can take the maximum score of both analysis (disease_model_max_norm) or a combination of them both (disease_model_avg_norm). In the Results and discussion section of the Phenodigm paper, it is described all the statistical analysis comparing the models to a number of gold standards. It is also concluded that the geometric mean of both analyses is the superior metric and should therefore be used as the score.

tskir commented:
Addressed the latest iteration of comments from @ireneisdoomed:

1. Disabled score normalisation: https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/745bc1de29944ce70eaec45dee93c88bd42894f7 — to be done by ETL instead.
2. As discussed, the assertion in `write_evidence_strings()`, as indeed most assertions, is not really intended to ever trigger, and is there as a cheap but important sanity check in case something goes catastrophically wrong.
3. Since the paper mentions that the mean score (`disease_model_avg_norm`) is superior to the maximum (`disease_model_max_norm`), I changed the parser so that it uses the former. I've also added a very helpful summary which Irene compiled earlier in this PR: https://github.com/opentargets/evidence_datasource_parsers/pull/80/commits/7d064d78d88daf9905eabbededb7b4ed8c61cbca.

I've also plotted a heatmap of disease_model_avg_norm (X) and disease_model_max_norm (Y) just to make sense how they are related:
![heatmap](https://user-images.githubusercontent.com/10669118/117687250-85603500-b1af-11eb-91e2-73e9ecaf19a9.png)

Since they are very well correlated, an important point is that changing from one to the other will definitely not break things drastically.

@ireneisdoomed @DSuveges I think everything's done now; if you also agree, then please can someone give this PR an approval :)

ireneisdoomed reviewed (approved):
All looks great.
